File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/pharmacy/shared/phrm-store-requisition-items.model.ts
import {
    FormBuilder,
    FormControl,
    FormGroup,
    Validators
} from '@angular/forms';
import { PHRMItemMasterModel } from "./phrm-item-master.model";
import { PHRMStoreRequisition } from "./phrm-store-requisition.model";

export class PHRMStoreRequisitionItems {
    public RequisitionItemId: number = 0;
    public ItemId: number = 0;
    public Quantity: number = 0;
    public ReceivedQuantity: number = 0;
    public PendingQuantity: number = 0;
    public RequisitionId: number = 0;
    public RequisitionItemStatus: string = ""
    public Remark: string = "";
    public AuthorizedBy: number = 0;
    public AuthorizedOn: string = "";
    public AuthorizedRemark: string = "";
    public CreatedBy: number = 0;
    public CreatedOn: string = "";
    public ReceivedBy: string = "";
    public RequisitionItemValidator: FormGroup = null;
    public BatchNo: string = "";

    ////to make the instance ItemMaster with new row
    public SelectedItem: PHRMItemMasterModel = null;

    //ItemName only for display purpose
    public ItemName: string = "";
    public Code: string = ""
    public Item: PHRMItemMasterModel = null;
    public Requisition: PHRMStoreRequisition = null;
    public CreatedByName: string = null;
    public DispatchedByName: string = null;
    StoreRackName: string = '';
    public RequiredQuantity: number = 0;
    AvailableQuantity: number = 0;
    DispatchedQuantity: number = 0;
    StoreId: any;
    //for display purpose ie Requisiting Store Name
    public RequestedSourceStore: string;

    constructor() {

        var _formBuilder = new FormBuilder();
        this.RequisitionItemValidator = _formBuilder.group({
            'ItemId': ['', Validators.compose([Validators.required])],
            'Quantity': ['', Validators.compose([this.positiveNumberValidator])],
        });
    }

    public IsDirty(fieldName): boolean {
        if (fieldName == undefined)
            return this.RequisitionItemValidator.dirty;
        else
            return this.RequisitionItemValidator.controls[fieldName].dirty;
    }


    public IsValid(): boolean { if (this.RequisitionItemValidator.valid) { return true; } else { return false; } } public IsValidCheck(fieldName, validator): boolean {
        if (fieldName == undefined) {
            return this.RequisitionItemValidator.valid;
        }
        else
            return !(this.RequisitionItemValidator.hasError(validator, fieldName));
    }

    positiveNumberValidator(control: FormControl): { [key: string]: boolean } {
        if (control) {
            if (control.value <= 0)
                return { 'invalidNumber': true };
        }

    }

}

File: Code/Components/DanpheEMR.ServerModel/PharmacyModels/PHRMStoreRequisitionItemsModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DanpheEMR.ServerModel
{
    public class PHRMStoreRequisitionItemsModel
    {
        [Key]
        public int RequisitionItemId { get; set; }
        public int ItemId { get; set; }
        public double? Quantity { get; set; }
        public double? ReceivedQuantity { get; set; }
        public double? PendingQuantity { get; set; }
        public int? RequisitionId { get; set; }
        public string RequisitionItemStatus { get; set; }
        public string Remark { get; set; }
        public int AuthorizedBy { get; set; }
        public DateTime? AuthorizedOn { get; set; }
        public string AuthorizedRemark { get; set; }
        public int? CreatedBy { get; set; }
        public DateTime? CreatedOn { get; set; }
        public double? CancelQuantity { get; set; }
        public int? CancelledBy { get; set; }
        public DateTime? CancelledOn { get; set; }
        public string CancelRemarks { get; set; }
        public virtual PHRMStoreRequisitionModel Requisition { get; set; }
        public virtual PHRMItemMasterModel Item { get; set; }
    }
}
